File: code-example-tests/python/pymongo/utils/comparison/expect.py
"""
Expect — fluent comparison API for PyMongo code-example tests.

Every comparison is delegated to the language-agnostic comparison kernel (a
native Go binary under ``tools/comparison-kernel/bin/``). The bridge serialises
each expected value into raw text the kernel can parse and each actual value
into a JSON-compatible payload, so behaviour stays consistent across language
suites.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, Optional, Set

from .kernel_bridge import _normalize_for_wire, kernel_compare


class ConfigurationError(ValueError):
    """Raised when the Expect API is configured incorrectly."""


def _looks_like_file_path(path_str: str) -> bool:
    """A heuristic check for whether *path_str* could plausibly be a file path:
    short, single-line, no obvious JSON-syntax characters at the start."""
    if not path_str or "\n" in path_str or "\r" in path_str:
        return False
    if len(path_str) > 1024:
        return False
    stripped = path_str.lstrip()
    if stripped.startswith("{") or stripped.startswith("["):
        return False
    return True


def _resolve_expected_file(path_str: str) -> Optional[Path]:
    """Locate an expected-output file by walking up to the package root and
    checking common locations (``examples/``, project root, raw path)."""
    if not _looks_like_file_path(path_str):
        return None
    p = Path(path_str)
    if p.is_absolute() and p.is_file():
        return p
    base = Path(__file__).resolve().parents[3]
    for candidate in (
        base / "examples" / path_str,
        base.parent / "examples" / path_str,
        base / path_str,
        Path(path_str),
    ):
        if candidate.is_file():
            return candidate
    return None


def _serialize_expected(expected: Any) -> str:
    """Turn an expected value into raw text the kernel can parse.

      * existing file path → file contents
      * string starting with ``{`` or ``[`` → as-is (MongoDB-syntax content)
      * everything else → BSON-normalised and JSON-serialised
    """
    if expected is None:
        return "null"
    if isinstance(expected, (str, Path)):
        str_val = str(expected)
        resolved = _resolve_expected_file(str_val)
        if resolved is not None:
            return resolved.read_text(encoding="utf-8")
        if isinstance(expected, str):
            stripped = str_val.lstrip()
            if stripped.startswith("{") or stripped.startswith("["):
                return str_val
        return json.dumps(str_val)
    return json.dumps(_normalize_for_wire(expected))


def _kernel_reports_parse_failure(response: Dict[str, Any]) -> bool:
    """Detect the kernel's "failed to parse expected content" condition so
    callers can fall back to plain text comparison for non-JSON files."""
    if response.get("error"):
        return True
    for err in response.get("errors") or []:
        if err.get("path") == "expected" and "failed to parse expected content" in (
            err.get("message") or ""
        ):
            return True
    return False


def _text_compare(expected_content: str, actual: Any) -> bool:
    """Plain-text equality fallback used when the kernel can't parse the
    expected content as JSON. Whitespace is normalised lightly so trailing
    newlines and CRLF differences don't trip simple text outputs."""
    if not isinstance(actual, str):
        return False
    norm = lambda s: s.replace("\r\n", "\n").rstrip()
    return norm(expected_content) == norm(actual)


def _build_error_message(response: Dict[str, Any]) -> str:
    """Format a kernel response's errors into a human-readable message."""
    if response.get("error"):
        return response["error"]
    errors = response.get("errors") or []
    if not errors:
        return "Comparison failed"
    parts = []
    for err in errors:
        path = err.get("path") or ""
        message = err.get("message") or ""
        parts.append(f"{path}: {message}" if path else message)
    return "\n".join(parts)


class Expect:
    """Fluent comparison API for PyMongo tests."""

    def __init__(self, actual: Any) -> None:
        self._actual = actual
        self._comparison_type: Optional[str] = None
        self._ignore_fields: Set[str] = set()
        self._has_sort_option = False
        self._using_should_match = False
        self._expected_for_resemble: Any = None

    @classmethod
    def that(cls, actual: Any) -> "Expect":
        return cls(actual)

    def with_ignored_fields(self, *fields: str) -> "Expect":
        if self._expected_for_resemble is not None:
            raise ConfigurationError(
                "with_ignored_fields() and should_resemble() are mutually exclusive. "
                "Use with_ignored_fields() with should_match() for exact comparisons, "
                "or use should_resemble() + with_schema() for schema-based validation."
            )
        self._ignore_fields = set(fields)
        return self

    def with_ordered_sort(self) -> "Expect":
        self._has_sort_option = True
        self._comparison_type = "ordered"
        return self

    def with_unordered_sort(self) -> "Expect":
        self._has_sort_option = True
        self._comparison_type = "unordered"
        return self

    def should_match(self, expected: Any) -> None:
        if self._expected_for_resemble is not None:
            raise ConfigurationError(
                "should_match() and should_resemble() are mutually exclusive. "
                "Use should_match() for exact comparisons or should_resemble() + with_schema() "
                "for schema-based validation, but not both."
            )
        self._using_should_match = True

        try:
            expected_content = _serialize_expected(expected)
        except Exception as exc:
            raise AssertionError(f"Failed to resolve expected value: {exc}")

        response = kernel_compare(
            expected_content,
            self._actual,
            comparison_type=self._comparison_type,
            ignore_field_values=sorted(self._ignore_fields) or None,
        )

        if response.get("isMatch"):
            return

        # Non-JSON expected content falls back to plain text comparison so
        # tests using raw multi-line text files keep working.
        if _kernel_reports_parse_failure(response) and _text_compare(
            expected_content, self._actual
        ):
            return

        raise AssertionError(_build_error_message(response))

    def should_resemble(self, expected: Any) -> "Expect":
        if self._using_should_match:
            raise ConfigurationError(
                "should_match() and should_resemble() are mutually exclusive. "
                "Use should_match() for exact comparisons or should_resemble() + with_schema() "
                "for schema-based validation, but not both."
            )
        if self._ignore_fields:
            raise ConfigurationError(
                "with_ignored_fields() and should_resemble() are mutually exclusive. "
                "Use with_ignored_fields() with should_match() for exact comparisons, "
                "or use should_resemble() + with_schema() for schema-based validation."
            )
        if self._has_sort_option:
            raise ConfigurationError(
                "with_ordered_sort()/with_unordered_sort() and should_resemble() are mutually exclusive. "
                "Sort options only apply to should_match() comparisons."
            )
        self._expected_for_resemble = expected
        return self

    def with_schema(self, schema: Dict[str, Any]) -> None:
        if self._expected_for_resemble is None:
            raise ConfigurationError(
                "with_schema() requires should_resemble() to be called first. "
                "Example: Expect.that(actual).should_resemble(expected).with_schema({...})"
            )
        if not isinstance(schema, dict) or not schema:
            raise ConfigurationError("with_schema() requires a schema dictionary")

        if "count" not in schema:
            raise ConfigurationError(
                "with_schema() requires a 'count' key specifying the expected number of documents"
            )
        count = schema["count"]
        if not isinstance(count, int) or isinstance(count, bool) or count < 0:
            raise ConfigurationError(
                "with_schema() requires 'count' to be a non-negative integer"
            )

        required_fields = schema.get("required_fields")
        if required_fields is not None and not isinstance(required_fields, list):
            raise ConfigurationError(
                "with_schema() 'required_fields' must be a list of field names"
            )

        field_values = schema.get("field_values")
        if field_values is not None and not isinstance(field_values, dict):
            raise ConfigurationError(
                "with_schema() 'field_values' must be a dictionary of key-value pairs"
            )

        if self._actual is None:
            raise AssertionError("Schema validation failed: actual output is None")
        if not isinstance(self._actual, (dict, list)):
            raise AssertionError(
                f"Schema validation failed: actual output must be a document or list of documents, "
                f"got {type(self._actual).__name__}"
            )

        try:
            expected_content = _serialize_expected(self._expected_for_resemble)
        except Exception as exc:
            raise AssertionError(f"Failed to resolve expected value: {exc}")

        kernel_schema = {
            "count": count,
            "requiredFields": required_fields or [],
            "fieldValues": _normalize_for_wire(field_values or {}),
        }

        response = kernel_compare(
            expected_content,
            self._actual,
            schema=kernel_schema,
        )

        if response.get("isMatch"):
            return

        raise AssertionError(
            f"Schema validation failed:\n{_build_error_message(response)}"
        )

File: code-example-tests/javascript/driver/utils/Expect.js
import { ComparisonEngine } from './comparison/comparisonEngine.js';
import path from 'path';

/**
 * Expect provides a fluent API for comparing MongoDB documentation example
 * outputs. Every comparison is delegated to the language-agnostic comparison
 * kernel (a native Go binary under tools/comparison-kernel/bin/), so behaviour
 * is identical across language suites.
 *
 * @example
 * // File comparison
 * Expect.that(actualResults).shouldMatch('expected-output.txt');
 *
 * // Object comparison with options
 * Expect.that(actualResults)
 *   .withOrderedSort()
 *   .withIgnoredFields('_id', 'timestamp')
 *   .shouldMatch(expectedResults);
 *
 * // Schema-based validation (results may vary, structure must hold)
 * Expect.that(actualResults)
 *   .shouldResemble(expectedOutput)
 *   .withSchema({
 *     count: 20,
 *     requiredFields: ['_id', 'title', 'year'],
 *     fieldValues: { year: 2012 },
 *   });
 */
class Expect {
  constructor(actual) {
    this._actual = actual;
    this._options = {
      arrayComparison: 'unordered',
      ignoreFields: [],
    };
    this._comparisonType = null;
    this._expectedForResemble = null;
    this._hasSortOption = false;
  }

  static that(actual) {
    return new Expect(actual);
  }

  withUnorderedSort() {
    this._hasSortOption = true;
    this._options.arrayComparison = 'unordered';
    return this;
  }

  withOrderedSort() {
    this._hasSortOption = true;
    this._options.arrayComparison = 'ordered';
    return this;
  }

  withIgnoredFields(...fields) {
    this._options.ignoreFields = fields;
    return this;
  }

  shouldMatch(expected) {
    if (this._comparisonType === 'schema') {
      throw new Error(
        'shouldMatch() cannot be used with shouldResemble() - they are mutually exclusive'
      );
    }
    this._comparisonType = 'exact';

    const result = ComparisonEngine.compare(expected, this._actual, {
      ...this._options,
      baseDir: path.resolve(__dirname, '../examples'),
    });

    if (!result.isMatch) {
      throw new Error(result.getErrorSummary());
    }
  }

  shouldResemble(expectedOutput) {
    if (this._comparisonType === 'exact') {
      throw new Error(
        'shouldResemble() cannot be used with shouldMatch() - they are mutually exclusive'
      );
    }
    if (this._options.ignoreFields.length > 0) {
      throw new Error(
        'withIgnoredFields() cannot be used with shouldResemble() - ' +
          'schema validation ignores field values by default'
      );
    }
    if (this._hasSortOption) {
      throw new Error(
        'withOrderedSort()/withUnorderedSort() cannot be used with shouldResemble() - ' +
          'sort options only apply to shouldMatch() comparisons'
      );
    }
    this._comparisonType = 'schema';
    this._expectedForResemble = expectedOutput;
    return this;
  }

  withSchema(schema) {
    if (this._comparisonType !== 'schema') {
      throw new Error(
        'withSchema() requires shouldResemble() to be called first'
      );
    }
    if (!schema || typeof schema !== 'object') {
      throw new Error('withSchema() requires a schema object');
    }
    if (
      typeof schema.count !== 'number' ||
      !Number.isInteger(schema.count) ||
      schema.count < 0
    ) {
      throw new Error(
        `withSchema() requires a non-negative count number, got ${schema.count}`
      );
    }
    if (
      schema.requiredFields !== undefined &&
      !Array.isArray(schema.requiredFields)
    ) {
      throw new Error(
        'withSchema() requiredFields must be an array of field names'
      );
    }
    if (
      schema.fieldValues !== undefined &&
      (typeof schema.fieldValues !== 'object' ||
        schema.fieldValues === null ||
        Array.isArray(schema.fieldValues))
    ) {
      throw new Error(
        'withSchema() fieldValues must be an object of key-value pairs'
      );
    }

    // Catch a common configuration mistake early: fieldValues should be a
    // subset of requiredFields. Otherwise a fieldValues entry can silently
    // pass for documents that lack the field altogether.
    const requiredFieldsSet = new Set(schema.requiredFields || []);
    const orphanFieldValues = Object.keys(schema.fieldValues || {}).filter(
      (k) => !requiredFieldsSet.has(k)
    );
    if (orphanFieldValues.length > 0) {
      const list = orphanFieldValues.map((f) => `'${f}'`).join(', ');
      throw new Error(
        `Schema validation configuration error: fieldValues contains field(s) [${list}] ` +
          `that are not in requiredFields. All fields in fieldValues must also be listed in requiredFields.`
      );
    }

    // Reject obviously-incompatible actual values up front so the user gets a
    // clearer error than the kernel's "document at index 0 is not an object".
    if (this._actual === null) {
      throw new Error('Schema validation failed: actual output is null');
    }
    if (this._actual === undefined || typeof this._actual !== 'object') {
      throw new Error(
        `Schema validation failed: actual output must be an array or document, got ${typeof this._actual}`
      );
    }

    const result = ComparisonEngine.compareWithSchema(
      this._expectedForResemble,
      this._actual,
      {
        count: schema.count,
        requiredFields: schema.requiredFields || [],
        fieldValues: schema.fieldValues || {},
      },
      { baseDir: path.resolve(__dirname, '../examples') }
    );

    if (!result.isMatch) {
      throw new Error(`Schema validation failed:\n${result.getErrorSummary()}`);
    }
  }
}

export default Expect;
export { Expect };
